core/unix/os.c

```diff
@@ -789,6 +796,106 @@ static init_fn_t
 #else
 /* If we're a normal shared object, then we override _init.
  */
+
+#    if defined(MUSL)
+/* We rely on PR_GET_AUXV to obtain a copy of auxvector and take one of the
+ * pointer entries as marker when searching through the address space.
+ *
+ * TODO: PR_GET_AUXV is relatively new (introduced in Linux 6.4), implement a
+ * fallback when it's unavailable.
+ * TODO i#7491: PR_GET_AUXV isn't correctly intercepted, thus executable-related
+ * auxvector entries cannot be used as target marker when searching, or
+ * applications linked to libdynamorio.so, like test client.gonative, will fail
+ * to execute when running under early injection. Switch to use a more-reliable
+ * address-related entry when the problem is resolved.
+ */
+static ELF_WORD
+get_auxv_value(ELF_WORD type)
+{
+    /* Currently no architecture defines more than 60 auxvector keys */
+    ELF_AUXV_TYPE auxv[64];
+
+    dynamorio_syscall(SYS_prctl, 5, PR_GET_AUXV, auxv, sizeof(auxv), 0, 0);
+
+    for (int i = 0; auxv[i].a_type != AT_NULL; i++) {
+        if (auxv[i].a_type == type)
+            return auxv[i].a_un.a_val;
+    }
+
+    ASSERT_NOT_REACHED();
+    return 0;
+}
+
+/* When entering the entry point, the stack layout looks like
+ *
+ *   sp => argc
+ *         argv
+ *         NULL (end of argv)
+ *         envp
+ *         NULL (end of envp)
+ *         auxv
+ *
+ * But the stack pointer has gone much farther towards the low-end of address
+ * space, thus it's hard to reliably locate the address of auxvector.
+ *
+ * search_auxvector() walks towards the higher address and locates one of the
+ * auxvector entries, then walks backwards and finds the beginning of auxvector.
+ */
+static void *
+search_auxvector(void *sp)
+{
+    ELF_WORD uid = get_auxv_value(AT_UID);
+
+    for (size_t offset = 0; offset < PAGE_SIZE * 64; offset += sizeof(ulong)) {
+        ELF_AUXV_TYPE *p = sp + offset;
+        ELF_AUXV_TYPE entry;
+
+        if (!d_r_safe_read(p, sizeof(ELF_AUXV_TYPE), &entry))
+            return NULL;
+
+        if (p->a_type == AT_UID && p->a_un.a_val == uid) {
```

Comment:
AT_UID is 11.  So this is just looking for 11 next to the uid: and the uid could be a small integer, even 0 for root: so this doesn't seem a very conclusive check.  I think it needs to look at a pointer value much less likely to be mistaken; ideally multiple values.